search/floydwarshall.go

```diff
@@ -156,38 +157,62 @@ func (p ShortestPaths) Between(u, v graph.Node) (path []graph.Node, weight float
 	}
 	path = []graph.Node{p.nodes[from]}
 	unique = true
+	seen := make([]int, len(p.nodes))
+	for i := range seen {
+		seen[i] = -1
+	}
+	seen[from] = 0
 	for from != to {
 		c := p.at(from, to)
 		if len(c) != 1 {
 			unique = false
 		}
-		from = c[rand.Intn(len(c))]
+		i := rand.Intn(len(c))
+		from = c[i]
+		if seen[from] >= 0 {
```

Comment:
I have a more general test for the non-productive path. I'll add that. It works with n=100, but is very slow with n=1000 (33s to complete cf. 0.005s for the n=10 case). If it becomes a problem, we can randomly permute the c slice and iterate through that slice, removing the chosen element. Can we leave this for a later PR when it is seen to be necessary?